src/openscm_twolayermodel/impulse_response_model.py

```diff
@@ -0,0 +1,327 @@
+            self._timestep_idx = 0
+
+        else:
+            self._timestep_idx += 1
+
+        if np.equal(self._timestep_idx, 0):
+            self._temp1_mag[self._timestep_idx] = 0.0
+            self._temp2_mag[self._timestep_idx] = 0.0
+            self._rndt_mag[self._timestep_idx] = 0.0
+
+        else:
+            self._temp1_mag[self._timestep_idx] = self._calculate_next_temp(
+                self._delta_t_mag,
+                self._temp1_mag[self._timestep_idx - 1],
+                self._q1_mag,
+                self._d1_mag,
+                self._erf_mag[self._timestep_idx - 1],
+            )
+
+            self._temp2_mag[self._timestep_idx] = self._calculate_next_temp(
+                self._delta_t_mag,
+                self._temp2_mag[self._timestep_idx - 1],
+                self._q2_mag,
+                self._d2_mag,
+                self._erf_mag[self._timestep_idx - 1],
+            )
+
+            self._rndt_mag[self._timestep_idx] = self._calculate_next_rndt(
+                self._temp1_mag[self._timestep_idx - 1],
+                self._temp2_mag[self._timestep_idx - 1],
+                self._erf_mag[self._timestep_idx - 1],
+                self._q1_mag,
+                self._q2_mag,
+                self._d1_mag,
+                self._d2_mag,
+                self._efficacy_mag,
+            )
+
+    @staticmethod
+    def _calculate_next_temp(delta_t, t, q, d, erf):
+        decay_factor = np.exp(-delta_t / d)
+        rise = erf * q * (1 - np.exp(-delta_t / d))
+
+        return t * decay_factor + rise
+
+    def _calculate_next_rndt(self, t1, t2, erf, q1, q2, d1, d2, efficacy):
+        two_layer_paras = self.get_two_layer_parameters()
+        lambda_0 = two_layer_paras["lambda_0"]
+
+        if np.equal(efficacy, 1):
+            efficacy_term = 0 * ur(self._erf_unit)
+        else:
+            eta = two_layer_paras["eta"]
+
+            C = two_layer_paras["du"] / (heat_capacity_water * density_water)
+            C_D = two_layer_paras["dl"] / (heat_capacity_water * density_water)
+
+            b_pt_1 = (lambda_0 + efficacy * eta) / C
+            b_pt_2 = eta / C_D
+            b = b_pt_1 + b_pt_2
+            b_star = b_pt_1 - b_pt_2
+            delta = b ** 2 - (4 * lambda_0 * eta) / (C * C_D)
+
+            phicoeff = C / (2 * efficacy * eta)
+            phi1 = phicoeff * (b_star - delta ** 0.5)
+            phi2 = phicoeff * (b_star + delta ** 0.5)
+
+            t1_h = t1 * ur(self._temp1_unit)
+            t2_h = t2 * ur(self._temp2_unit)
+            efficacy_term = (
+                eta * (efficacy - 1) * ((1 - phi1) * t1_h - (1 - phi2) * t2_h)
+            )
+
+            if str(efficacy_term.units) != "watt / meter ** 2":
+                raise AssertionError("units should have come out as W/m^2")
+
+        out = erf - lambda_0.magnitude * (t1 + t2) - efficacy_term.magnitude
+
+        return out
+
+    def _get_run_output_tss(self, ts_base):
```

Comment:
"tss" is just plural for "ts" as timeseries?

Reply:
yep, nastily named but internal so maybe ok?

Reply:
sure! maybe add a comment in the base class?

Reply:
done